1071 allow selection of download file internal structure

juzen2003 opened the pull request:
- Fixes #1071 
- Were any Django, import pipeline, table_schema, or dictionary files modified? N
  - If YES:
    - Database used: opus3_test_210331
    - All Django tests pass: Y
    - FLAKE8 run on modified code: Y
- Were any JavaScript or CSS files modified? Y
  - If YES:
    - JSHINT run on all affected files: Y (cart.js)
    - Tested on Chrome, Firefox, Safari, Edge, and iOS: Y (No iOS)
      (Remember to test all browser sizes)
    - Tested for race conditions (with delayed API calls if applicable): NA

Description of changes:
- Add a "Hierarchical" checkbox below the "Download Options" in the cart pane. 
    - If checkbox is selected, the downloaded zip file will have all files stored with hierarchy tree.
    - If checkbox is not selected, in the zip file:
        - files with the same basename + same logical path (duplicated) will only be stored once.
        - files with the same basename + different logical paths (not duplicated) will be stored within its own hierarchy tree.
- API tests for /api/download/<opusid>.zip & /__cart/download.json are added.
- For local testing purpose, the default selected files (0 byte dummy files) for these opus ids are stored in Shared OPUS dropbox:  
    - co-iss-n1460973661
    - co-iss-n1481265970
    - co-iss-n1462840881
  
Known problems:
N/A

juzen2003 committed:
Modify api_create_download to have an option "hierarchical_struct" to
determine if files stored in zip file should have hierarchy tree.
- In views.py (line 650-667), loop through all files to create a
  dictionary keyed by basenames. Each key has a list of paths pointing
  to itself. If there are multiple paths for a key, then it means these
  paths are not duplicated and need to be stored with hierarchy tree in
  the zip file.
- In views.py (line 695-699), add the logic to make sure if
  hierarchical_struct is True or there are multiple paths for the same
  file basename, we store files with hierarchy tree in the zip file.

juzen2003 committed:
Add a checkbox to download data pane in the cart. The checkbox will be
used to determine if files should store in the zip file with hierarchy
tree.
- In views.py (656), get hierarchical_struct from the parameter passed
  in to __cart/download.json. If it's 1, then files should store in the
  zip file with hierarchy tree.
- In cart.js (line 349-353), pass in hierarchical value to api call
  based on whether the hierarchical checkbox is selected.
- In cart.html (line 16-19, 40-46), add hierarchical checkbox to some
  possible locations, will have to modify and update this later.

juzen2003 committed:
Add API tests for /__cart/download.json.
- In views.py (line 669), modify the logic to make sure we don't store
  duplicated files with hierarchy tree.
- In api_test_helper.py (line 171-187), create _run_zip_equal to compare
  the filenames in the zip file created by /__cart/download.json.
- In test_cart_api.py (line 2379-2447), add API tests for
  /__cart/download.json to test out the followings:
    - download single opus id and store with/without hierarchy tree
    - download multiple opus ids with duplicated files and store
      with/without hierarchy tree
    - download multiple opus ids without duplicated files and store
      with/without hierarchy tree

juzen2003 committed:
Update the location of hierarchical checkbox, put it under "Download
Options" in the download pane.
- In cart.html (line 36-39), put hierarchical checkbox below "Download
  Options".
- In opus.css (line 3227-3231), add styling to make sure when "Download
  Options" is gone, we create some space to between the checkbox and
  download buttons.
- In opus.css (line 2917-2921), add styling to have hierarchical checkbox
  to fit in at the bottom of "Download Options".

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 1071_allow_selection_of_download_file_internal_structure

juzen2003 committed:
Add API tests for /api/download/<opusid>.zip.
- In api_test_helper.py (line 175-183), update _run_zip_equal to parse
  the response differently based on different API calls (__cart/download.json
  or __api/download/).
- In test_cart_api.py (line 2449-2463), add tests for /api/download/<opusid>.zip.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -516,6 +516,8 @@ def api_create_download(request, opus_id=None):
         or: [__]api/download/(?P<opus_id>[-\w]+).zip
     Arguments: types=<PRODUCT_TYPES>
                urlonly=1 (optional) means to not zip the actual data products
+               hierarchical=0 (optional) means files in zip are not stored with
```
Since hierarchical=0 is the default if nothing is specified, this should instead say what happens if you specify hierarchical=1 (like urlonly=1)

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -3217,6 +3223,12 @@ when wrapping into the 2nd line */
         display: none !important;
     }
+
+    /* Make sure when "Download Options" is gone, we create some space to between
```
Remove "to"

rfrenchseti commented on opus/application/test_api/api_test_helper.py:
```diff
@@ -166,3 +167,25 @@ def _run_csv_equal_file(self, url, exp_file):
+        if '__cart/download.json' in url:
+            jdata = json.loads(response.content)
+            file = jdata['filename']
+            filename = file[file.rindex('/'):]
```
Since we aren't guaranteed what the user has set for TAR_FILE_URL_PATH, how about just stripping off TAR_FILE_URL_PATH (e.g. replace(TAR_FILE_URL_PATH, '') instead of looking for the last "/"?

rfrenchseti commented on opus/application/test_api/api_test_helper.py:
```diff
@@ -166,3 +167,25 @@ def _run_csv_equal_file(self, url, exp_file):
+        print(url)
+        response = self._get_response(url)
+        self.assertEqual(response.status_code, 200)
+        if '__cart/download.json' in url:
```
Rather than check explicitly for download,json and download/ how about passing in a flag that says what kind of response you expect? Then it's still useable if we change the API in the future.

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -2373,22 +2373,97 @@ def test__api_cart_datacsv_multi(self):
             ######### /__cart/download.json: API TESTS #########
             ####################################################
+    def test__api_cart_download_single_no_hierarchical(self):
+        "[test_cart_api.py] /__cart/download.json: single opus id  & no hierarchical"
```
Extra space before &

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -2373,22 +2373,97 @@ def test__api_cart_datacsv_multi(self):
             ####################################################
+    def test__api_cart_download_single_no_hierarchical(self):
+        "[test_cart_api.py] /__cart/download.json: single opus id  & no hierarchical"
+        url = '/__cart/add.json?opusid=co-iss-n1462840881&reqno=456'
```
Should call reset.json at the beginning of each test like the other existing tests do

rfrenchseti commented on opus/application/test_api/test_cart_api.py:
```diff
@@ -2373,22 +2373,97 @@ def test__api_cart_datacsv_multi(self):
+        self._run_json_equal(url, expected)
+        url = '/__cart/download.json?types=coiss_raw,coiss_calib,browse_full&hierarchical=0'
+        expected = ['N1462840881_1.IMG', 'N1462840881_1.LBL', 'prefix2.fmt', 'tlmtab.fmt', 'N1462840881_1_CALIB.IMG', 'N1462840881_1_CALIB.LBL', 'N1462840881_1_full.png', 'manifest.csv', 'data.csv', 'urls.txt']
+        self._run_zip_equal(url, expected)
```
All of these tests leave .zip files behind in the TAR_FILE_PATH. They need to clean up after themselves by removing the zip file.

rfrenchseti reviewed (changes requested):
Now that I look at the wording, I think it would make more sense to have the checkbox say "Flat zip file structure" and be checked by default. We have plenty of room for that with this UI layout. You can keep the flag name and whatnot and just change the meaning of the text.

Also maybe the line that says "Select which product types..." should have a little more visual contrast. Perhaps make it a heavier character weight? Something that makes it look more like a heading, which it really is.

juzen2003 committed:
Change the name of "Hierarchical" checkbox to "Flat zip file structure",
and it's checked by default. Make modifications based on review feedback.
- In cart.html (line 36-38), change the name of "Hierarchical" checkbox
  to "Flat zip file structure", and also change the class name to
  "op-download-flat-zip-file-structure".
- In views.py (line 519), update the function description of
  api_create_download.
- In opus.css (line 2923-2926), add font weight to make "Select all
  product types" and "Deselect all product types" more visible.
- In cart.js (line 350-351), update the logic to determine the value for
  hierarchical param passed into the API.
- In api_test_helper.py (line 172), add a new param response_type for
  _run_zip_equal, it will be used to determine how to parse the response.
- In api_test_helper.py (line 193-196), remove the .zip stored under
  settings.TAR_FILE_PATH after the test.
- In test_cart_api.py, call reset.json for all /api/download/<opusid>.zip
  and /__cart/download.json tests.

juzen2003 committed:
Update the string comparison to use "==" instead of "is" because "is"
means "are exactly the same object" not "have equal contents".

rfrenchseti commented on opus/application/test_api/api_test_helper.py:
```diff
@@ -166,3 +168,29 @@ def _run_csv_equal_file(self, url, exp_file):
+        elif response_type == 'binary':
+            binary_stream = BytesIO(response.content)
+            zip_file = zipfile.ZipFile(binary_stream, mode='r')
+        resp = zip_file.namelist()
```
If you're going to use a series of "if...elif" then there needs to be something to catch the last clause, like an "else: assert False" or something. Otherwise if you pass in the wrong argument it will simply crash with an undefined variable. Alternatively you can just replace the "elif" with a plain "else".

rfrenchseti commented on opus/application/test_api/api_test_helper.py:
```diff
@@ -166,3 +168,29 @@ def _run_csv_equal_file(self, url, exp_file):
+            jdata = json.loads(response.content)
+            file = jdata['filename']
+            filename = file.replace(settings.TAR_FILE_URL_PATH, '')
+            path = settings.TAR_FILE_PATH + filename
```
Can't you just say file.replace(settings.TAR_FILE_URL_PATH, settings.TAR_FILE_PATH)?

rfrenchseti commented on opus/application/static_media/css/opus.css:
```diff
@@ -2914,6 +2914,17 @@ when wrapping into the 2nd line */
+}
+
+/* Make "Select all product types" and "Deselect all product types" more visible */
+.op-cart-select-all-btn, .op-cart-deselect-all-btn {
```
I meant the line above the buttons, not the buttons themselves.

juzen2003 committed:
Remove the font weight for select all and deselect all product types
buttons in the download pane.

juzen2003 committed:
Update based on github review.
- In api_test_helper.py (line 179-180), update the logic to get the path
  of zip file.
- In api_test_helper.py(line 182), change the condition block to else to
  avoid crash when not passing in 'json'/'binary' in response_type.
